### PR TITLE
Feature grpc auth server for publisher revenue

### PR DESCRIPTION
**Pull request by ghun131**
# Motivation / Background

Publisher Revenue need an API to check authenticated token or not

*This Pull Request has been created because:*
Check token authentication using gRPC https://thepressingly.atlassian.net/browse/PINET-354 

# Detail

- [x] gRPC server method at Lago to verify auth token
- [x] Unit test for this method

# Additional information

### How to integrate
- Copy `auth_token.proto` file content
- Generate new code using the command similar to `grpc_tools_ruby_protoc`. Take note that this command is for `ruby`, with different language, you'll have a different command
- At local pass `LAGO_GRPC_URL="localhost:50051"` to turn on gRPC server
```shell
grpc_tools_ruby_protoc -I grpc/publisher_revenue/protos --ruby_out=packs/publisher_portal/lib --grpc_out=output_code_path/lib path_to_proto_file/auth_token.proto
```

# Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.

### CODE REVIEW

File: app/services/grpc/server/auth_token_service.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+require 'grpc'
+require 'auth_token_services_pb'
+
+module Grpc
+  module Server
+    class AuthTokenService < Auth::Token::Service
+      def verify(token_req, _unused_call)
+        if token_req.token && decoded_token(token_req.token) && valid_token?
```
Comment (TronTram) [inline app/services/grpc/server/auth_token_service.rb:10]: why not reuse the AuthenticableUser module?

Comment (ghun131) [reply app/services/grpc/server/auth_token_service.rb:10]: That's good point. I got trouble with `renew_token` method

Comment (ghun131) [reply app/services/grpc/server/auth_token_service.rb:10]: So when I add `include AuthenticableUser` to `AuthTokenService` class. The error when running the server is this
```bash
/Users/ghun131/Code/Work/PiNet/lago-api-3/app/controllers/concerns/authenticable_user.rb:7:in `block in <module:AuthenticableUser>': undefined method `before_action' for class Grpc::Server::AuthTokenService (NoMethodError)

    before_action :renew_token
    ^^^^^^^^^^^^^
	from /Users/ghun131/.rvm/gems/ruby-3.3.0/gems/activesupport-7.0.8.1/lib/active_support/concern.rb:136:in `class_eval'
	from /Users/ghun131/.rvm/gems/ruby-3.3.0/gems/activesupport-7.0.8.1/lib/active_support/concern.rb:136:in `append_features'
	from /Users/ghun131/Code/Work/PiNet/lago-api-3/app/services/grpc/server/auth_token_service.rb:9:in `include'
	from /Users/ghun131/Code/Work/PiNet/lago-api-3/app/services/grpc/server/auth_token_service.rb:9:in `<class:AuthTokenService>'
	from /Users/ghun131/Code/Work/PiNet/lago-api-3/app/services/grpc/server/auth_token_service.rb:8:in `<module:Server>'
	from /Users/ghun131/Code/Work/PiNet/lago-api-3/app/services/grpc/server/auth_token_service.rb:7:in `<module:Grpc>'
	from /Users/ghun131/Code/Work/PiNet/lago-api-3/app/services/grpc/server/auth_token_service.rb:6:in `<main>'
	from config.ru:8:in `require_relative'
	from config.ru:8:in `block in <main>'
	from /Users/ghun131/.rvm/gems/ruby-3.3.0/gems/rack-2.2.8.1/lib/rack/builder.rb:116:in `eval'
```

Comment (ghun131) [reply app/services/grpc/server/auth_token_service.rb:10]: It seems that I can reuse `current_user` method from `AuthenticableUser` so I must reimplement `valid_token?` method in some way.

**Review by TronTram — APPROVED** [review]
lgtm

**Comment by TronTram** [comment]
i have a suggestion:
```ruby 
# app/services/token_service.rb

# frozen_string_literal: true

module TokenService
  def self.decode_token(token)
    JWT.decode(token, ENV['SECRET_KEY_BASE'], true, decode_options)
  rescue JWT::DecodeError => e
    Rails.logger.error("Error decoding token: #{e}")
    raise e if e.is_a?(JWT::ExpiredSignature) || Rails.env.development?
  end

  def self.valid_token?(decoded_token)
    payload = decoded_token.reduce({}, :merge)
    Time.now.to_i <= payload['exp']
  end

  def self.decode_options
    {
      algorithm: 'HS256',
    }
  end
end
```
In  AuthenticableUser

```ruby
# frozen_string_literal: true

require 'token_service'

module AuthenticableUser
  extend ActiveSupport::Concern

  included do
    before_action :renew_token
  end

  private

  # Reference to TokenService for decoding token
  def current_user
    @current_user ||= User.find_by(id: payload_data['sub']) if token && decoded_token && valid_token?
  end

  # Reference to TokenService for extracting token from request headers
  def token
    @token ||= request.headers['Authorization'].to_s.split(' ').last
  end

  # Reference to TokenService for decoding the token
  def decoded_token
    @decoded_token ||= TokenService.decode_token(token)
  rescue JWT::DecodeError => e
    raise e if e.is_a?(JWT::ExpiredSignature) || Rails.env.development?
  end

  # Reference to TokenService for checking token validity
  def valid_token?
    TokenService.valid_token?(@decoded_token)
  end

  # Reference to TokenService for getting payload data from the decoded token
  def payload_data
    @payload_data ||= @decoded_token.reduce({}, :merge)
  end

  # This method remains here to handle renewing the token
  def renew_token
    return unless current_user

    result = UsersService.new.new_token(current_user)
    response.set_header('x-lago-token', result.token) if result.success?
  end
end
```

In AuthTokenService
```ruby
# frozen_string_literal: true

require 'grpc'
require 'auth_token_services_pb'
require 'token_service' # Ensure this file requires the TokenService module

module Grpc
  module Server
    class AuthTokenService < Auth::Token::Service
      def verify(token_req, _unused_call)
        if token_req.token && valid_token?(token_req.token)
          Auth::AuthResponse.new(status: "succeed")
        else
          Auth::AuthResponse.new(status: "failed")
        end
      end

      private

      def valid_token?(token)
        decoded_token = TokenService.decode_token(token)
        return false unless decoded_token
     
        TokenService.valid_token?(decoded_token)
      rescue JWT::DecodeError => e
        Rails.logger.error("Error decoding token: #{e}")
        raise e if e.is_a?(JWT::ExpiredSignature) || Rails.env.development?
      end
    end
  end
end
```
TODO: update test, remove duplicate methods

**Pull request merged**